src/Menu/useMenu.ts

```diff
@@ -0,0 +1,281 @@
+import React from 'react';
+import { FocusContext, useFocusContext, useFocusItem } from '../useFocusContext';
+import { useIds } from '../useId';
+import { useOutsideClick } from '../useOutsideClick';
+import { useUpdateEffect } from '../useUpdateEffect';
+import { normalizeEventKey } from '../utils';
+
+export type ElementProps<Element> = React.DetailedHTMLProps<React.HTMLAttributes<Element>, Element>;
+
+interface MenuContext {
+   isOpen: boolean;
+   focusContext: FocusContext;
+   menuId: string;
+   menuButtonId: string;
+   menuButtonRef: React.MutableRefObject<HTMLButtonElement | null>;
+   menuRef: React.MutableRefObject<HTMLUListElement | null>;
+   openAndFocusFirstItem: () => void;
+   openAndFocusLastItem: () => void;
+   close: () => void;
+}
+
+export const MenuContext = React.createContext<MenuContext | null>(null);
+
+export function useMenuContext(): MenuContext {
+   const context = React.useContext(MenuContext);
+   if (context == null) {
+      throw new Error('Cannot use useMenuContext without Menu provider');
+   }
+   return context;
+}
+
+export type MenuProps = React.PropsWithChildren<{
+   id?: string;
+   className?: string;
+   defaultIsOpen?: boolean;
+}>;
+
+export function useMenu({ id, defaultIsOpen = false }: MenuProps): MenuContext {
+   const [isOpen, setIsOpen] = React.useState(defaultIsOpen);
+   const menuButtonRef = React.useRef<HTMLButtonElement>(null);
+   const menuRef = React.useRef<HTMLUListElement>(null);
+   const [menuButtonId, menuId] = useIds(id, `menu-button`, `menu-list`);
+   const focusContext = useFocusContext();
+
+   const close = React.useCallback(() => {
+      setIsOpen(false);
+   }, []);
+
+   const openAndFocusFirstItem = React.useCallback(() => {
+      setIsOpen(true);
+      focusContext.focusFirst();
+   }, [focusContext]);
+
+   const openAndFocusLastItem = React.useCallback(() => {
+      setIsOpen(true);
+      focusContext.focusLast();
+   }, [focusContext]);
+
+   useOutsideClick(menuRef, (event) => {
+      if (isOpen && !menuButtonRef.current?.contains(event.target as HTMLElement)) {
+         close();
+      }
+   });
+
+   const context = React.useMemo((): MenuContext => {
+      return {
+         isOpen,
+         focusContext,
+         menuButtonId,
+         menuId,
+         menuButtonRef,
+         menuRef,
+         openAndFocusFirstItem,
+         openAndFocusLastItem,
+         close,
+      };
+   }, [
+      menuButtonId,
+      menuId,
+      menuButtonRef,
+      menuRef,
+      openAndFocusFirstItem,
+      openAndFocusFirstItem,
+      isOpen,
+      focusContext,
+      close,
+   ]);
+
+   return context;
+}
+
+export type MenuListProps = React.HTMLAttributes<HTMLUListElement>;
+
+export type MenuListInternalProps = ElementProps<HTMLUListElement> & {
```

Comment:
Looks like this isn't used outside this file, so I think it doesn't need to be exported.

Reply:
Yes this is type isn't meant to be exposed 👍